### DISCUSSION TITLE
UnifiedMap: Collect map entry points and desired behavior

### DISCUSSION THREAD

**Discussion by moving-bits** [cgeo/cgeo#13121] [Q&A]
We have several entry points to our map (eg: cache list, address search, single cache map etc.), and each entry point has slightly different behavior (eg: zoom level, map center, map title etc.). Currently a so called "map state" is used to initialize the map for different scenarios. A "map state" is basically a bunch of attributes passed to the map, and the map needs to interpret what to make of it when some attributes are filled, others are not. This leads to different behavior between GM and OSM maps in several cases. 

For implementing `UnifiedMap` I would like to harmonize the behavior and to name the different scenarios a bit more explicitly. For this I would like to collect information: Which entry points do we have, what data will be passed to the map in what case, and how should the map behave in the scenario given?

I've started collecting a few bits on a new wiki page ([Map usages](https://github.com/cgeo/cgeo/wiki/Map-usages)) already - but that's more for documenting the behavior finally agreed upon & implemented. 

Let's collect and discuss the different scenarios first in this discussion thread. As a starter, I've added a few basic scenarios below.
- scenario: Name / short description of the scenario
- entry point: How to start map in this scenario?
- additional data passed to map: What data needs to be passed to the map for it to be able to start this scenario?
- map actions: How should the map be initialized, what actions should be called on starting in this scenario?

Some of the things that differ between the scenarios are:
- zoom level
- map center
- set / reset target
- map title
- what to display (caches, waypoints, ...)
- enable / disable live-mode
- ...

If you have any remarks, corrections etc. for a given scenario, please answer as reply to the specific scenario.
Feel free to add additional scenarios (as separate answers).

**Comment by moving-bits** [comment]
### scenario 
default map

### entry point
start via bottom navigation

### additional data passed to map
no additional data given

### map actions
- display map with last coords and last zoom level for default map
- live-mode = (use current setting)
- title = "map" or "live map" (depending on live-mode status)
- reset target

**Comment by moving-bits** [comment]
### scenario 
cache map

### entry point
start from cache popup/detail page via navigate => map

### additional data passed to map
geocode

### map actions
- load cache
- display map at zoom level xxx
- center map to cache coords
- set cache as target
- live-mode = disabled
- title = cache name

**Comment by moving-bits** [comment]
### scenario 
address map

### entry point
start from address search via map icon

### additional data passed to map
coords

### map actions
- set coords as center
- use last zoom level for default map
- live-mode = (use current setting)
- title = "map" or "live map", depending on live-mode status
- reset target

**Comment by moving-bits** [comment]
## scenario
cache list map

## entry point
start from cache list via map icon

## additional data passed to map
set of geocodes (or just list id? - to be discussed)

## map actions
- adjust zoom level and bounds to display all caches in given list
- reset target
- title = list name

## to be discussed: list switching from map window
Currently we take a list of caches that are part of a list, and call the map with a list of geocodes. The map then does not know any longer which list this is coming from. One thing being missed because of this is to be able to switch cache lists directly in map view. I see two possible alternatives for this:

- instead of passing a list of geocodes to the map, pass the info (map type) "cache list" and the list's id to the map, and let the map handle cache loading etc. - possible drawback: which filter should get applied to the map - map filter or last list filter?
- or pass a bunch of geocodes as today, but allow changing the list anyway, reloading the bunch of geocodes in this case.

**Reply by ztNFny** [reply]
I would very much like this to change to loading the actual list instead of just a bunch of points. If list is filtered the filter should be applied to map as well. Switching the selected list should then also be possible from map view.
I see "list" and "map" just as 2 views of the exact same thing and would like them handled as such. (In the same way the reverse is currently true for viewing live map and then selecting "view as list")

**Reply by moving-bits** [reply]
Reminder: If something changes in the list, map needs to get triggered that there is an update. E. g.:
- open cache popup from map, remove cache from currently displayed list => cache needs to be remove from map
- store live cache (not yet in list) into list => cache needs to stay on map

**Reply by ztNFny** [reply]
Switching back and forth between list and cache map it'd be nice if zoom level and position would be kept (instead of zooming to show all caches). Usecase: I'm viewing a list as map, then accidentally click "back" once to often. When viewing the list as map again I now have to find the correct zoom level and view again.

Maybe this could be combined with bottom nav somehow? Bottom nav remembers view state already. Maaaybeeee we could get rid of the "show liust as map" button as well, integrating it with bottom nav (but of course many open questions there)

**Comment by ztNFny** [comment]
### scenario 
waypoint map

### entry point
waypoints tab on a cache

### additional data passed to map
no additional data given

### map actions
- set coords as center
- use last zoom level for default map
- live-mode = off
- title = waypoint name
- set as target
